### ISSUE TITLE
Try to automatically detect konnector issues

### ISSUE THREAD

**Issue by Chocobozzz** [cozy-labs/konnectors#375]
See https://github.com/cozy-labs/konnectors/issues/363

The idea is to use (real world) user konnector exceptions/bad status code to detect if a konnector doesn't work anymore. The purpose is to help developers to maintain their konnectors.

**Comment by bnjbvr** [comment]
Nice, that's interesting. For what it's worth, that's also an issue for Kresus : when Weboob specific bank modules aren't up to date anymore, we can have connectivity issues between the bank and the Cozy. The people at Budgea, the company behind Weboob, have had an idea of a "weather map" of connections: they just retrieve, from each user, whether the connection was successful or not. Then, they display the ratio `successful connections / total connections` on a public weather-like map, for each bank.

**Comment by bnjbvr** [comment]
Another idea I had was to have a small test server, that one could run on their own personal server. This server would just try to connect to the provider's website, for a given single (or more) konnectors, using the CLI. This server would then send its result to a centralized server that could then send alerts to maintainers.

So for instance, let's say I'm the maintainer of X, and Alice is the maintainer of Y. I can have a server connecting to X and Y with my credentials every night. Whenever I get into trouble connecting to one of these services, my testing server sends a ping to a centralized server : and when X is not working, I receive an email; when Y is not working, Alice receives an email.

Of course, emails shall not be sent for each ping, so there should be some mitigation, or some smart system determining whether it's a real issue or just a fluke in data.

**Comment by ZeHiro** [comment]
I think it is hard to detect a failed konnector : 
- detecting a failure in the connection process can be easy (but this means strengthening the process to detect connection errors)
- detecting a change in the page layout is much harder (is cheerio returning no value because there is no data to parse, or is it because the page changed?)

**Comment by nicofrand** [comment]
Maybe we could detect some parsing errors with cheerio by first selecting a higher DOM node.
Ex: to retrieve the list of my bills I use 
```js
let bills = $("#bills li");
```
Then, if `bills.length` (just an example, we actually check `bills.fetched.length`) returns 0, we do not know if this is an error or there was simply no result.

If  Instead, we select the container : 
```js
let container = $("#bills");
```
and then each element : 

```js
container.find("li").each()
```

If `container.length` returns 0, the layout probably changed.

**Comment by Phyks** [comment]
Another idea, to notify the user something has broken would be to add a field to specify time to warn after last event.

Typically, if I connect my Online account, and I pay my server monthly there, I may want to know if it has been more than a month without retrieving any bill. Then, I know that there may be an error with this konnector and can check manually.

I am not sure how feasible / easy it would be to implement it though, I am quite new to using Konnectors.

**Comment by nicofrand** [comment]
That might be another solution, but that would not detect a broken konnector the first time you launch it.

**Comment by Phyks** [comment]
Sure. That said, if a connector was notifying when it does not return anything at first launch and if a user could set a time after which he wants a notification if no data has been retrieved, it could help him check that konnectors work.

I agree this is not a very technical solution, but I think it would improve UX quite a lot. At the moment, I have many konnectors fetching monthly bills for instance. If I could have a notification in case none of them has retrieved anything within one month (plus some security margin), I would know it is most likely broken, and could debug / open an issue, without having to check periodically that everything works fine.

Another feature that would be useful I think would be to have a link to see last imported items directly.

Let me know if you think these ideas would deserve their own issue.

**Comment by nicofrand** [comment]
At first launch you can also launch the konnector manually and see if some bills are fetched.
Maybe it is even done by konnectors itself when you setup a monthly check, I do not know. Maybe @ZeHiro or @doubleface know ?

There already is link to open the directory in Files.

**Comment by ZeHiro** [comment]
@nicofrand : I already do a regular check on my own konnectors. Except for doctolib, hopefully I do not have to go that much to the doctor :smile:

**Comment by nicofrand** [comment]
@ZeHiro yep, but my idea would alert the user while you correct your connector ;).
My idea would suppose that every connector using cheerio would do it that way though :(.

**Comment by ZeHiro** [comment]
@nicofrand 
There could be a more "disgusting" way, at least for monthly/regularly basis konnectors : 
- Fetch all imported data from the db (with appropriate parameters)
- if there is already imported information and the konnector finds nothing (as all the konnector fetch all information), raise an error.

**Comment by nicofrand** [comment]
Yes, but that would not work for the first-time use.
But again, both solutions can be implemented.

**Comment by ZeHiro** [comment]
The problem is that sometimes for the first time use, there is nothing to import (but it is normal)

**Comment by Phyks** [comment]
About the manual check, sure it is easy to do it manually at the moment, but providing easy access and notifications to automate the process would improve a lot the UX and give an impression of reliability I think.

Concerning the first time use, which potentially has nothing to import, if there is just a warning which is displayed, it won't impact much the behavior, and the user can do a manual check and see that everything is normal. Something like "this konnector did not retrieve any data during initial import, there may be a problem with the konnector".
